### PR TITLE
CUDA build

### PR DESCRIPTION
**Pull request by izahn**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

This is a cleaned up PR that adds a CUDA build. It is based on the work and feedback in https://github.com/conda-forge/tensorflow-feedstock/pull/118 

The main issue is that I couldn't get a CUDA version to build with system libraries. I finally gave up and unset `TF_SYSTEM_LIBS` for the CUDA builds only (other builds should still use system libraries as before). It would be even more awesome if we could get CUDA builds and use conda system libraries. I tried hard but couldn't figure out how to do it, and I think a CUDA build will be valuable, even if we can't get system libraries working.

### CODE REVIEW

File: recipe/build.sh [hunk 1]
```diff
@@ -106,9 +97,89 @@ export TF_DOWNLOAD_CLANG=0
...
+    --copt=-mtune=haswell
+    --copt=-ftree-vectorize
+    --copt=-fPIC
+    --copt=-fstack-protector-strong
+    --copt=-O2
+    --cxxopt=-fvisibility-inlines-hidden
+    --cxxopt=-fmessage-length=0
+    --linkopt=-zrelro
+    --linkopt=-znow
+    --linkopt=-L${PREFIX}/lib
```
Comment (xhochy) [inline recipe/build.sh:148]: These things should all go into the bazel toolchain. The scripts for that also use the default conda-forge flags automatically. The ones listed here are only a subset and need to be manually maintained.

File: recipe/build.sh [hunk 2]
```diff
@@ -106,9 +97,89 @@ export TF_DOWNLOAD_CLANG=0
...
+    --action_env=PYTHON_LIB_PATH=${SP_DIR}
+    --python_path=${PYTHON}
+    --define=PREFIX=$PREFIX
+    --copt=-DNO_CONSTEXPR_FOR_YOU=1
+    --host_copt=-DNO_CONSTEXPR_FOR_YOU=1
+    --define=LIBDIR=$PREFIX/lib
+    --define=INCLUDEDIR=$PREFIX/include"
+
+    ## Doesn't work with CUDA builds, don't know why
+    export TF_SYSTEM_LIBS=""
```
Comment (xhochy) [inline recipe/build.sh:165]: What's the error you're getting?

**Review by xhochy — COMMENTED** [review]
Have a look at the bazel toolchain and the scripts that generate it. This was the groundbreaking part for me to get the CPU builds working.

File: recipe/build.sh (hunk 1 above)
Comment (izahn) [reply recipe/build.sh:148]: got, it, thanks!

File: recipe/build.sh (hunk 2 above)
Comment (izahn) [reply recipe/build.sh:165]: It wasn't finding headers, e.g. I would get

```
Execution platform: @local_execution_config_platform//:platform
In file included from tensorflow/core/platform/protobuf.cc:16:
./tensorflow/core/platform/protobuf.h:28:10: fatal error: google/protobuf/io/coded_stream.h: No such file or directory
   28 | #include "google/protobuf/io/coded_stream.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

I believe this was a consequence of not using `custom_tookchain`. Now that I'm back on track with that I no longer get these errors.

**Comment by izahn** [comment]
> Have a look at the bazel toolchain and the scripts that generate it. This was the groundbreaking part for me to get the CPU builds working.

It turns out that the `CFLAGS` set in https://github.com/conda-forge/nvcc-feedstock/blob/master/recipe/install_nvcc.sh were the cause of the `The include path '/usr/local/cuda/include' references a path outside of the execution root.` errors I was getting, and that was the thing that led me to avoid the `custom_toolchain`. I think I have a work-around for that, and am building locally with `custom_toolchain` now. If it works locally I'll push these changes.

**Comment by izahn** [comment]
I think I'm making progress, but hit another snag I don't understand:
```
ERROR: /home/conda/feedstock_root/build_artifacts/tensorflow-split_1622495488587/work/tensorflow/compiler/mlir/lite/quantization/BUILD:138:20: Linking of rule '//tensorflow/compiler/mlir/lite/quantization:op_quant_spec_getters_gen' failed (Exit 1): crosstool_wrapper_driver_is_not_gcc failed: error executing command 
  (cd /home/conda/.cache/bazel/_bazel_conda/f7f47ae89e86280f733605a1378b0008/execroot/org_tensorflow && \
  exec env - \
    PATH=/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622495488587/work:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622495488587/_build_env/bin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622495488587/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/bin:/opt/conda/condabin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622495488587/_build_env:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622495488587/_build_env/bin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622495488587/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622495488587/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/bin:/opt/conda/bin:/opt/conda/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/conda/bin:/usr/local/cuda/bin \
    PWD=/proc/self/cwd \
  external/local_config_cuda/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc @bazel-out/host/bin/tensorflow/compiler/mlir/lite/quantization/op_quant_spec_getters_gen-2.params)
Execution platform: @local_execution_config_platform//:platform
/usr/bin/ld: cannot find -lz
collect2: error: ld returned 1 exit status
INFO: Elapsed time: 5284.940s, Critical Path: 197.14s
INFO: 11007 processes: 8270 internal, 2737 local.
FAILED: Build did NOT complete successfully
```
Any ideas on this one? Full log at https://gist.github.com/izahn/47c950b53ffca4e8f68818b67538d495 in case that helps.

**Comment by izahn** [comment]
Still having problems with this:

```
  [11,311 / 15,254] Compiling tensorflow/core/kernels/relu_op_gpu.cu.cc; 111s local ... (4 actions running)
ERROR: /home/conda/feedstock_root/build_artifacts/tensorflow-split_1622561486085/work/tensorflow/tools/proto_text/BUILD:31:10: Linking of rule '//tensorflow/tools/proto_text:gen_proto_text_functions' failed (Exit 1): crosstool_wrapper_driver_is_not_gcc failed: error executing command 
  (cd /home/conda/.cache/bazel/_bazel_conda/8cea2436bc4316f162f1b5b3058eb302/execroot/org_tensorflow && \
  exec env - \
    PATH=/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622561486085/work:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622561486085/_build_env/bin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622561486085/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/bin:/opt/conda/condabin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622561486085/_build_env:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622561486085/_build_env/bin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622561486085/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622561486085/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/bin:/opt/conda/bin:/opt/conda/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/conda/bin:/usr/local/cuda/bin \
    PWD=/proc/self/cwd \
  external/local_config_cuda/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc @bazel-out/host/bin/tensorflow/tools/proto_text/gen_proto_text_functions-2.params)
Execution platform: @local_execution_config_platform//:platform
/usr/bin/ld: cannot find -lprotobuf
/usr/bin/ld: cannot find -lsnappy
/usr/bin/ld: cannot find -lprotobuf
collect2: error: ld returned 1 exit status
ERROR: /home/conda/feedstock_root/build_artifacts/tensorflow-split_1622561486085/work/tensorflow/core/framework/BUILD:1261:31 Linking of rule '//tensorflow/tools/proto_text:gen_proto_text_functions' failed (Exit 1): crosstool_wrapper_driver_is_not_gcc failed: error executing command 
  (cd /home/conda/.cache/bazel/_bazel_conda/8cea2436bc4316f162f1b5b3058eb302/execroot/org_tensorflow && \
  exec env - \
    PATH=/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622561486085/work:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622561486085/_build_env/bin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622561486085/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/bin:/opt/conda/condabin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622561486085/_build_env:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622561486085/_build_env/bin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622561486085/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622561486085/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/bin:/opt/conda/bin:/opt/conda/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/conda/bin:/usr/local/cuda/bin \
    PWD=/proc/self/cwd \
  external/local_config_cuda/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc @bazel-out/host/bin/tensorflow/tools/proto_text/gen_proto_text_functions-2.params)
Execution platform: @local_execution_config_platform//:platform
INFO: Elapsed time: 17383.472s, Critical Path: 1300.39s
INFO: 11640 processes: 6284 internal, 5356 local.
FAILED: Build did NOT complete successfully
```

Full log at https://gist.github.com/izahn/47c950b53ffca4e8f68818b67538d495

bazel is really killing me here, I still don't understand it.

File: recipe/build.sh [hunk 3]
```diff
@@ -110,6 +115,41 @@ export TF_CONFIGURE_IOS=0
...
+    fi
+    
+    export TF_NEED_CUDA=1
+    export TF_CUDA_VERSION="${cuda_compiler_version}"
+    export TF_CUDNN_VERSION="${cudnn}"
+    export TF_CUDA_CLANG=0
+    export TF_DOWNLOAD_CLANG=0
+    export TF_NEED_TENSORRT=0
+    export TF_NCCL_VERSION=""
+    BUILD_OPTS="${BUILD_OPTS} --config=cuda --linkopt=-L${PREFIX}/lib --define=LIBDIR=${PREFIX}/lib"
```
Comment (hmaarrfk) [inline recipe/build.sh:148]:
```suggestion
    BUILD_OPTS="${BUILD_OPTS} -s --config=cuda --linkopt=-L${PREFIX}/lib --define=LIBDIR=${PREFIX}/lib"
```

Apparently `-s` makes it print out the compilation commands.

Comment (hmaarrfk) [reply recipe/build.sh:148]: you might even want to try changing `--linkopt` to `--copt` or even `--host_linkopt`

Comment (izahn) [reply recipe/build.sh:148]: Thanks, yeah that is very verbose, but didn't really give any more information about what does wrong.

Comment (hmaarrfk) [reply recipe/build.sh:148]: how long does it take to get to failure?

Comment (izahn) [reply recipe/build.sh:148]: Time to failure varies, because the order isn't always the same. Can be anywhere from 30 minutes to several hours.

Comment (hmaarrfk) [reply recipe/build.sh:148]: sigh, this makes it harder to debug.

**Comment by izahn** [comment]
> Have a look at the bazel toolchain and the scripts that generate it. This was the groundbreaking part for me to get the CPU builds working.

I notice
```
WARNING: option '--config=cuda' (source command line options) was expanded and now overrides the explicit option --crosstool_top=//custom_toolchain:toolchain with --crosstool_top=@local_config_cuda//crosstool:toolchain
```
in the logs, and suspect that might be a problem, if not the problem, causing my builds to fail.

File: recipe/build.sh (hunk 3 above)
Comment (izahn) [reply recipe/build.sh:148]: > you might even want to try changing `--linkopt` to `--copt` or even `--host_linkopt`

I think `--copt=-L${PREFIX}/lib` might have done the trick! I'm almost afraid to get my hopes up at this point, but we are safely past the place where the build failed last time. Fingers crossed!

Comment (izahn) [reply recipe/build.sh:148]: Nope, it fails later with

```
[12,897 / 29,765] Compiling tensorflow/lite/toco/graph_transformations/resolve_multiply_by_zero.cc [for host]; 3s local ... (4 actions running)
ERROR: /home/conda/feedstock_root/build_artifacts/tensorflow-split_1622642338317/work/tensorflow/tools/proto_text/BUILD:31:10: Linking of rule '//tensorflow/tools/proto_text:gen_proto_text_functions' failed (Exit 1): crosstool_wrapper_driver_is_not_gcc failed: error executing command 
  (cd /home/conda/.cache/bazel/_bazel_conda/fc5b6f8a245ea6c7bfa068d002f44f78/execroot/org_tensorflow && \
  exec env - \
    PATH=/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622642338317/work:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622642338317/_build_env/bin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622642338317/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/bin:/opt/conda/condabin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622642338317/_build_env:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622642338317/_build_env/bin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622642338317/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622642338317/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/bin:/opt/conda/bin:/opt/conda/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/conda/bin:/usr/local/cuda/bin \
    PWD=/proc/self/cwd \
  external/local_config_cuda/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc @bazel-out/host/bin/tensorflow/tools/proto_text/gen_proto_text_functions-2.params)
Execution platform: @local_execution_config_platform//:platform
/usr/bin/ld: cannot find -lprotobuf
/usr/bin/ld: cannot find -lsnappy
/usr/bin/ld: cannot find -lprotobuf
collect2: error: ld returned 1 exit status
ERROR: /home/conda/feedstock_root/build_artifacts/tensorflow-split_1622642338317/work/tensorflow/core/framework/BUILD:1261:31 Linking of rule '//tensorflow/tools/proto_text:gen_proto_text_functions' failed (Exit 1): crosstool_wrapper_driver_is_not_gcc failed: error executing command 
  (cd /home/conda/.cache/bazel/_bazel_conda/fc5b6f8a245ea6c7bfa068d002f44f78/execroot/org_tensorflow && \
  exec env - \
    PATH=/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622642338317/work:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622642338317/_build_env/bin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622642338317/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/bin:/opt/conda/condabin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622642338317/_build_env:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622642338317/_build_env/bin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622642338317/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1622642338317/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/bin:/opt/conda/bin:/opt/conda/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/conda/bin:/usr/local/cuda/bin \
    PWD=/proc/self/cwd \
  external/local_config_cuda/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc @bazel-out/host/bin/tensorflow/tools/proto_text/gen_proto_text_functions-2.params)
Execution platform: @local_execution_config_platform//:platform
INFO: Elapsed time: 13126.052s, Critical Path: 444.51s
INFO: 13232 processes: 8181 internal, 5051 local.
FAILED: Build did NOT complete successfully
```

full log at https://gist.github.com/izahn/42712a8437361cb40aa5aed129f05155

**Comment by izahn** [comment]
I've taken this as far as I can for the time being. I've pushed my latest changes, and I think its in pretty good shape except for the linker issue. If someone has time and interest in picking it up that would be great, otherwise I'll try to come back to this at some later date. Thanks to everyone who offered their time and advice, I appreciate your help!

**Comment by hmaarrfk** [comment]
I'm finding it strange that there are references to the use of compilers in `/usr/bin`.

For one, the `/usr/bin/ld: cannot find -lz`.

It should be using the conda-forge `ld` compiler.

I'm expecting it to be something like
```
${HOST_PREFIX}/bin/x86_64-conda-linux-gnu-ld
```

File: recipe/build.sh [hunk 4]
```diff
@@ -110,6 +102,50 @@ export TF_CONFIGURE_IOS=0
 sed -i -e "/PROTOBUF_INCLUDE_PATH/c\ " .bazelrc
 sed -i -e "/PREFIX/c\ " .bazelrc
 
+if [[ ${cuda_compiler_version} != "None" ]]; then
+
+    export GCC_HOST_COMPILER_PATH="${GCC}"
```
Comment (hmaarrfk) [inline recipe/build.sh:107]:
```suggestion
    export GCC_HOST_COMPILER_PATH="${GCC}"
    export GCC_HOST_COMPILER_PREFIX="$(dirname ${GCC})"
    # No obvious way to specify a different ld, so just link it to conda-forge's
    ln -s ${LD} $(dirname ${LD})/ld
```

Comment (hmaarrfk) [reply recipe/build.sh:107]: `$(dirname ${LD})` and `$(dirname ${GCC})` should be the same thing.

File: recipe/build.sh [hunk 5]
```diff
@@ -110,6 +102,50 @@ export TF_CONFIGURE_IOS=0
...
+    elif [[ ${cuda_compiler_version} == 11.1 ]]; then
+        export TF_CUDA_COMPUTE_CAPABILITIES=5.2,5.3,6.0,6.1,6.2,7.0,7.2,7.5,8.0,8.6
+    elif [[ ${cuda_compiler_version} == 11.2 ]]; then
+        export TF_CUDA_COMPUTE_CAPABILITIES=5.2,5.3,6.0,6.1,6.2,7.0,7.2,7.5,8.0,8.6
+    else
+        echo "unsupported cuda version."
+        exit 1
+    fi
+    
+    BUILD_OPTS="${BUILD_OPTS} --config=cuda --copt=-L${PREFIX}/lib --define=LIBDIR=${PREFIX}/lib --define=PREFIX=$PREFIX"
```
Comment (hmaarrfk) [inline recipe/build.sh:133]: Maybe try adding the `-L` flag to `CROSSTOOL.template`???

**Comment by hmaarrfk** [comment]
For reference here is Anaconda's recipe https://github.com/AnacondaRecipes/tensorflow_recipes/blob/master/tensorflow-base-gpu/build.sh

**Comment by izahn** [comment]
> For reference here is Anaconda's recipe https://github.com/AnacondaRecipes/tensorflow_recipes/blob/master/tensorflow-base-gpu/build.sh

Yes, I have a working conda-forge build based on this at https://github.com/izahn/tensorflow-feedstock/tree/anaconda. I even built it, you can test it out from https://anaconda.org/izahn/repo . The main issues with it are a) it doesn't use conda system libraries, and b) the anaconda and conda-forge recipes have diverged considerably, making it difficult to merge the anaconda approach here.

**Comment by hmaarrfk** [comment]
Right, i noticed that they diverged. I'm wondering if we can bring them back together.

**Comment by hmaarrfk** [comment]
I mostly left it here since it was hard for me to find it.

**Comment by hmaarrfk** [comment]
whoa! did you find it?

**Comment by izahn** [comment]
> whoa! did you find it?

Kind of! I gave up on getting `TF_SYSTEM_LIBS` working with the CUDA build, and just let `bazel` build everything, like they do in https://github.com/AnacondaRecipes/tensorflow_recipes/blob/master/tensorflow-base-gpu. I also used the patches from there. This results in extremely long build times (all I have available for building is a 4 year  old thinkpad, takes about two days...), and a huge package at the end of it all. But... it works!

The log from my most recent completed build is at https://gist.github.com/izahn/b194405de3595d460199cbc3bbce0c74, and the packages are at https://anaconda.org/izahn/repo. One of the `libtensorflow` tests failed, but the main `tensorflow-base` package works, and I "tested" it by running https://pypi.org/project/ai-benchmark/.

All in all I think I'm getting very close to something that I would be comfortable merging here. It needs investigation and fix for the failing `libtensorflow` test, and maybe a little more code cleanup, but I think we're close!

**Comment by izahn** [comment]
The linux builds are working, both with and without `cuda`.  Unfortunately it looks like I broke the osx builds, despite my efforts not to change anything with them.

Several build logs are available for review, and the corresponding artifacts are available at https://anaconda.org/izahn/repo 

- [cuda111py38hdd42111_1.log](https://gist.github.com/izahn/5406d1b51f185b9706706ab03528da98)
- [cuda111py39h769e6be_1.log ](https://gist.github.com/izahn/41d566b473938c914890a3a9dbaba046)
- [cpu_py39h5fcbf9e_1.log ](https://gist.github.com/izahn/5406d1b51f185b9706706ab03528da98)
- [cpu_py38h2c3d97f_1.log](https://gist.github.com/izahn/7d0eb52b977d6c6a864849f7c60dd52f)

Given the complexity of this build, and the fact that I had to resort to using a totally different build process for `cuda` packages, I'm not sure it makes sense to keep trying to do all this in one recipe. It may be simpler to split the `cuda` builds out into a separate `tensorflow-gpu` recipe, as Anaconda does with their tensorflow packages. Please let me know if you have thoughts or suggestions about that.

**Comment by 183amir** [comment]
> Given the complexity of this build, and the fact that I had to resort to using a totally different build process for `cuda` packages, I'm not sure it makes sense to keep trying to do all this in one recipe. It may be simpler to split the `cuda` builds out into a separate `tensorflow-gpu` recipe, as Anaconda does with their tensorflow packages. Please let me know if you have thoughts or suggestions about that.

We can have a `cuda` or `gpu`  branch here to track GPU recipes.

**Comment by hmaarrfk** [comment]
Sorry for chiming in so late. This just came back on my radar.

I think we should avoid having two branches.

I think that we can likely try to see if the OSX builds are just broken as is, or if we did something special in this recipe to break them.

I think I can try to build on the CIs the OSX part to see what broke.

**Comment by izahn** [comment]
If we go the direction of this PR we effectively have two completely different builds. The cuda enabled recipe in this PR is basically just the Anaconda one, while the non-cuda build has diverged considerably from the anaconda recipe. Because of this I think a separate branch makes sense. Alternatively we can keep trying to get a cuda enabled build based on the conda-forge recipe here rather than based on the Anaconda recipe. I tried pretty hard to do that but never could get it working. The only thing that came from that effort was that I learned to hate bazel.

**Comment by h-vetinari** [comment]
Thanks for pushing this forward!

I don't like the divergence in recipes TBH, but I understand the realities of the situation (and your hate for bazel). I haven't looked in detail, but is there an argument against using the CUDA-enabled recipe to also build the non-CUDA version?

Also, in the hope of enriching our understanding of the different approaches - have you seen the work from [open-ce](https://github.com/open-ce/tensorflow-feedstock/tree/main/recipe) on packaging tensorflow for conda?

At the time I [commented](https://github.com/conda-forge/tensorflow-feedstock/pull/110#issuecomment-758592173):
> Just discovered the following work by open-ce: https://github.com/open-ce/tensorflow-feedstock/blob/master/recipe
> [...]
>
> Since both the license of the [feedstock](https://github.com/open-ce/tensorflow-feedstock/blob/master/LICENSE) and the open-ce [project](https://github.com/open-ce/open-ce/blob/master/LICENSE) are Apache-2, there shouldn't be an issue with including some of this work here?

and @jayfurmanek responded:
> 
> 
> Absolutely no problem. TensorFlow is a beast, we'd love to collaborate. Feel free to pull in any of the work we did (perhaps some credit or a link in a readme or comment would be nice). A couple of things to note:
> 
> * The recipes in open-ce back to `defaults`, not to forge, so there could be some unforeseen issues related to that, especially since forge uses a newer toolchain.
> 
> * We enable TensorRT, cudnn for the cuda variant. I know forge has cuda packages now, at least so thats's good.
> 
> * Our recipe creates a separate `libtensorflow` package as well that includes the extra C and C++ API libs and headers. There are likely other ways of providing those, but this is the best way we've found to include all of the needed headers and to avoid rebuilding everything, just for the C and C++ libs.

**Comment by xhochy** [comment]
> Thanks for pushing this forward!
> 
> 
> 
> I don't like the divergence in recipes TBH, but I understand the realities of the situation (and your hate for bazel). I haven't looked in detail, but is there an argument against using the CUDA-enabled recipe to also build the non-CUDA version?

Not really as the CUDA-enabled one uses the system compilers and not the conda-forge ones. We need the c-f ones on OSX independent of the target and for cross-compiling on Linux. The latter is not used yet for Tensorflow but has been working really well with other bazel-based projects using the `bazel-toolchain` package. The conda-forge Tensorflow build is actually just taking the Anaconda one for `osx-64` to all platforms, there is divergence but coming from the same origin.

I haven't looked into any issues here as I actually don't need CUDA for my current use cases but I would expect that we would get it working with a limited amount of patching. The main issue here is not Bazel but just the extreme way of Tensorflow of doing all things on their own weird way. Bazel isn't a straight forward to use tool, it requires a lot of upfront knowledge but the issues here cannot be really blamed on it 😉

> Also, in the hope of enriching our understanding of the different approaches - have you seen the work from [open-ce](https://github.com/open-ce/tensorflow-feedstock/tree/main/recipe) on packaging tensorflow for conda?
> 

I had a brief look at that when doing the last iteration on building Tensorflow and it also helped and gave me some helpful pointers but as I didn't enable CUDA at that point, I didn't use anything from there.

File: .ci_support/linux_64_cuda_compiler_version11.1cudnn8python3.8.____cpython.yaml [hunk 6]
```diff
@@ -1,19 +1,25 @@
...
 - conda-forge main
+cuda_compiler:
+- nvcc
+cuda_compiler_version:
+- '11.1'
+cudnn:
+- '8'
 cxx_compiler:
-- clangxx
+- gxx
```
Comment (hmaarrfk) [inline .ci_support/linux_64_cuda_compiler_version11.1cudnn8python3.8.____cpython.yaml:18]: Could this be the issue with OSX?

**Comment by wolfv** [comment]
@xhochy are you planning on working on this? I'd love to help out as we need a CUDA-enabled build.

**Comment by xhochy** [comment]
> @xhochy are you planning on working on this? I'd love to help out as we need a CUDA-enabled build.

Not really, I would have a look from time to time into the errors that come up but I don't need CUDA currently.

**Comment by izahn** [comment]
The basic problem I've had when trying to make a cuda-enabled build based on the conda-forge recipe is that bazel stubbornly refuses to use the host system cuda headers. The error usually looks like
```
The include path '/usr/local/cuda/include' references a path outside of the execution root
```
@xhochy if you have any ideas about that it would be appreciated!

**Comment by isuruf** [comment]
> bazel stubbornly refuses to use the host system cuda headers.

One option is to just copy the headers to `${PREFIX}/include` and delete them at the end. (And also remove -I/usr/local/cuda/include` from CFLAGS, CPPFLAGS, CXXFLAGS)

**Comment by izahn** [comment]
> > bazel stubbornly refuses to use the host system cuda headers.
> 
> One option is to just copy the headers to `${PREFIX}/include` and delete them at the end. (And also remove -I/usr/local/cuda/include` from CFLAGS, CPPFLAGS, CXXFLAGS)

I'll be more than happy if someone wants to try this, but I've already spent more time than I care to admit working on this. I have an alternative that works (see https://github.com/conda-forge/tensorflow-feedstock/issues/134 ) and I don't plan to spend more time on this approach. I'm closing this PR, but I'll leave my branch up in case someone else wants to give it a shot.

**Pull request closed without merging**